Enable Intel® VTune™ Profiler's Instrumentation and Tracing Technology APIs (ITT) to PyTorch

jingxu10 opened the pull request:
More detailed description of benefits can be found at #41001. This is Intel's counterpart of NVidia’s NVTX (https://pytorch.org/docs/stable/autograd.html#torch.autograd.profiler.emit_nvtx).

ITT is a functionality for labeling trace data during application execution across different Intel tools.
For integrating Intel(R) VTune Profiler into Kineto, ITT needs to be integrated into PyTorch first. It works with both standalone VTune Profiler [(https://www.intel.com/content/www/us/en/developer/tools/oneapi/vtune-profiler.html](https://www.intel.com/content/www/us/en/developer/tools/oneapi/vtune-profiler.html)) and Kineto-integrated VTune functionality in the future.
It works for both Intel CPU and Intel XPU devices.

Pitch
Add VTune Profiler's ITT API function calls to annotate PyTorch ops, as well as developer customized code scopes on CPU, like NVTX for NVidia GPU.

This PR rebases the code changes at https://github.com/pytorch/pytorch/pull/61335 to the latest master branch.

Usage example:
```
with torch.autograd.profiler.emit_itt():
    for i in range(10):
        torch.profiler.itt.range_push('step_{}'.format(i))
        model(input)
        torch.profiler.itt.range_pop()
```

cc @ilia-cher @robieta @chaekit @gdankel @bitfort @ngimel @orionr @nbcsm @guotuofeng @guyang3532 @gaoteng-git

jingxu10 committed:
Enable VTune Profiler ITT functionality

fix flair8 and mypy issues

replaced tabs to spaces in torch/csrc/itt_wrapper.cpp

replaced tabs to spaces in torch/csrc/autograd/profiler_itt.cpp

rebase to c9ccb850

fine tune

fine tune according to comments

amend according to comments

update ittapi

resume itt func in onednn

fine tune docstring for emit_itt class

add range context to torch.profiler.itt

fix lint warnings

fix lint issues

fix lint issue

fix test_public_bindings failure

VitalyFedyunin commented:
MyPy failures seem legit

jingxu10 commented:
Hi @VitalyFedyunin, fixed the mypy issues.

gdankel commented:
In general looks fine, but can you avoid making changes to the legacy profiler? We should not add new features to that.

jingxu10 commented:
Hi @gdankel , the edit in legacy files are for the following:
1. add `struct ITTStubs`. Currently `struct CUDAStubs` is in `torch/csrc/autograd/profiler_legacy.h:35`.
2. add `ITT` to `enum class C10_API_ENUM ProfilerState` in `torch/csrc/autograd/profiler_legacy.h:400`.

It looks like the edit to the `ProfilerState` is anyway required.
I can move the `struct ITTStubs` to `torch/csrc/autograd/profiler_kineto.h`, but `struct CUDAStubs` will be left in `profiler_legacy.h`. I think it might be better to keep these 2 in the same file. If it doesn't matter to make them in separate files I'll move it to `profiler_kineto.h`.
Alternatively, how about keeping the definition of `struct ITTStubs` in `torch/csrc/autograd/profiler_legacy.h`, but removing all code changes in `torch/csrc/autograd/profiler_legacy.cpp`?

jingxu10 commented:
Hi @gdankel  @VitalyFedyunin , any updates?

robieta commented:
+ @malfet for CMake changes. I can review the profiler specific bits.

malfet commented on CMakeLists.txt:
```diff
@@ -291,6 +291,7 @@ if(NOT USE_XNNPACK AND CMAKE_VERSION VERSION_LESS ${XNNPACK_MIN_CMAKE_VER})
   message(WARNING "USE_XNNPACK is set to OFF. XNNPACK requires CMake version ${XNNPACK_MIN_CMAKE_VER} or greater.")
 endif()
 option(USE_ZMQ "Use ZMQ" OFF)
+option(USE_ITT "Use Intel(R) VTune Profiler ITT functionality" ON)
```
This should only be enabled for X86 CPU, isn't it?
```suggestion
cmake_dependent_option(USE_ITT "Use Intel(R) VTune Profiler ITT functionality" ON "CPU_INTEL" OFF)
```

jingxu10 replied:
this will also work for Intel GPU (XPU)

malfet replied:
But it would not work on ARM64, would it?

jingxu10 replied:
Correct. Not expect it working on ARM64. I'll bound it to IntelCPU

malfet commented on cmake/Modules/FindITT.cmake:
```diff
@@ -0,0 +1,28 @@
+SET(ITT_INCLUDE_DIR)
+SET(ITT_LIBRARIES)
+
+IF(NOT CMAKE_BUILD_TYPE)
+  SET(CMAKE_BUILD_TYPE "Release" CACHE STRING
+      "Choose the type of build, options are: Debug Release"
+      FORCE)
+ENDIF(NOT CMAKE_BUILD_TYPE)
```
Can you explain why this is needed?
```suggestion
```

jingxu10 replied:
This is to build ITTAPI into release mode.

malfet replied:
But it should be build in the same mode as rest of the build, isn't it?

jingxu10 replied:
Yes, it follows the rest of the build.

malfet commented on tools/build_variables.bzl:
```diff
@@ -150,6 +150,7 @@ libtorch_profiler_sources = [
     "torch/csrc/profiler/api.cpp",
     "torch/csrc/profiler/kineto_shim.cpp",
     "torch/csrc/profiler/nvtx_observer.cpp",
+    "torch/csrc/profiler/itt_observer.cpp",
```
Hmm, will this compile on non-x86 cpus?

jingxu10 replied:
Yes, I think so. This file will be compiled on all platforms.
It checks if ITT is enabled in runtime.
  TORCH_CHECK(
      torch::profiler::impl::ittStubs()->enabled(),
      "Can't use ITT profiler - PyTorch was compiled without ITT");

malfet commented on cmake/Modules/FindITT.cmake:
```diff
@@ -0,0 +1,28 @@
+
+SET(ITT_ROOT "${PROJECT_SOURCE_DIR}/third_party/ittapi")
+FIND_PATH(ITT_INCLUDE_DIR ittnotify.h PATHS ${ITT_ROOT} PATH_SUFFIXES include)
+IF (ITT_INCLUDE_DIR)
+  ADD_SUBDIRECTORY(${ITT_ROOT})
+  SET(ITT_LIBRARIES ittnotify)
+  SET(ITT_FOUND ON)
+ENDIF (ITT_INCLUDE_DIR)
```
This looks like a really old pattern, can you please use `add_library(ITT` and `target_include_directories` instead?
This way, ITT includes would be applied only to objects depending on it (and this dependency should be private, i.e. those headers do not need to be installed as part of libtorch, do they?)

jingxu10 replied:
Sure.
No, ITT headers are not necessary to be exposed out. They can be private.

jingxu10 replied:
HI @malfet , just would like to double confirm with you. ittapi already has a `CMakeLists.txt` containing `add_library` at https://github.com/intel/ittapi/blob/master/CMakeLists.txt. What we do in the `FindITT.cmake` is to invoke this `CMakeLists.txt`. Should we use `add_subdirectory` in the `FindITT.cmake` or `add_library` in this case?

malfet commented on torch/csrc/profiler/api.h:
```diff
@@ -139,6 +141,30 @@ struct TORCH_API CUDAStubs {
+// ----------------------------------------------------------------------------
+// -- ITT ---------------------------------------------------------------------
+// ----------------------------------------------------------------------------
+using ITTEventStub = std::shared_ptr<CUevent_st>;
+
+struct TORCH_API ITTStubs {
+  virtual void record(int* device, ITTEventStub* event, int64_t* cpu_ns)
+      const = 0;
+  virtual float elapsed(const ITTEventStub* event, const ITTEventStub* event2)
+      const = 0;
+  virtual void ittMark(const char* name) const = 0;
+  virtual void ittRangePush(const char* name) const = 0;
+  virtual void ittRangePop() const = 0;
+  virtual bool enabled() const {
+    return false;
+  }
+  virtual void onEachDevice(std::function<void(int)> op) const = 0;
+  virtual void synchronize() const = 0;
+  virtual ~ITTStubs();
+};
```
This looks like a word-by-word copy of `CUDAStubs` implemented just a few lines above, i.e.
https://github.com/pytorch/pytorch/blob/214951bc6bfff104483cde623289d04fa0b18588/torch/csrc/profiler/api.h#L121-L125

Why not rename `CUDAStubs` into something agnostic (say `ProfilerStubs`) and use them for both CUDA and ITT?

jingxu10 replied:
Do you mean dispatch to CUDA or ITT inside each function?

jingxu10 replied:
@malfet `CUDAStubs` is exposed in header files. Renaming it will cause changes to frontend API. Is it OK to do this?
./torch/include/torch/csrc/profiler/api.h:123:struct TORCH_API CUDAStubs {
./torch/include/torch/csrc/profiler/api.h:136:  virtual ~CUDAStubs();
./torch/include/torch/csrc/profiler/api.h:139:TORCH_API void registerCUDAMethods(CUDAStubs* stubs);
./torch/include/torch/csrc/profiler/api.h:140:TORCH_API const CUDAStubs* cudaStubs();

robieta commented on torch/autograd/profiler.py:
```diff
@@ -486,6 +486,70 @@ def _call_end_callbacks_on_future(self, fut: Future[Any]) -> Future[Any]:
+        ...     model(x)
+
+    """
+    def __init__(self, enabled=True, record_shapes=False):
```
When would one ever pass `enabled=False`?

jingxu10 replied:
For instance, when user is debugging his/her script, he/she can simply set this value to False to disable ITT, rather than remove the with clause. The same usage as emit_nvtx.

robieta commented on torch/autograd/profiler.py:
```diff
@@ -486,6 +486,70 @@ def _call_end_callbacks_on_future(self, fut: Future[Any]) -> Future[Any]:
+    def __enter__(self):
+        if not self.enabled:
+            return
+        if self.entered:
```
`enableProfiler` is already not reentrant. And this warning won't trigger on something like:
```
def f(x):
  with enable_itt():
    model(x)
    
with enable_itt():
  f(x)
```
which is how I would expect this sort of thing to crop up most often. Maybe it makes sense to move this message to C++?

jingxu10 replied:
This is also to align with `emit_nvtx`.

robieta commented on torch/autograd/profiler.py:
```diff
@@ -486,6 +486,70 @@ def _call_end_callbacks_on_future(self, fut: Future[Any]) -> Future[Any]:
         return profiled_future
 
 
+class emit_itt(object):
```
`torch.autograd.profiler` is a deprecated namespace. New development should be in `torch.profiler.`

jingxu10 replied:
As `emit_nvtx` is still in `torch.autograd.profiler`, `emit_itt` aligns with `emit_nvtx`.

robieta commented on torch/autograd/profiler.py:
```diff
@@ -486,6 +486,70 @@ def _call_end_callbacks_on_future(self, fut: Future[Any]) -> Future[Any]:
         return profiled_future
 
 
+class emit_itt(object):
```
Nit: `EmitITT`. `profile` is cased like a function for historic reasons, but new code doesn't need to.

jingxu10 replied:
I don't follow here. Should we change `emit_itt` to another word? This aligns with `emit_nvtx`.

robieta commented on torch/itt.py:
```diff
@@ -0,0 +1,39 @@
+try:
```
This doesn't seem to warrant its own `torch.foo` namespace. Why don't we make these static methods on the context manager? E.g. something like:
```
def f():
  torch.profiler.ITT.mark("bbb")

with torch.profiler.ITT() as p:
  model(x)
  p.mark("aaa")
  f()
```
That way you could access it either from the active profiler or just from the module.

jingxu10 replied:
Does this mean merging `emit_itt` into `torch.profiler.itt`?

robieta commented on torch/csrc/itt_wrapper.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef PROFILER_ITT_H
```
Why not just put these in `torch/csrc/profiler/api.h`? Also, should be `torch::profiler` namespace

jingxu10 replied:
Because `itt_mark`, `itt_range_push` and `itt_range_pop` are implemented in `torch/csrc/itt_wrapper.cpp`, to align with it, declaration of them are put in `torch/csrc/itt_wrapper.h`

robieta commented on torch/csrc/itt_wrapper.cpp:
```diff
@@ -0,0 +1,21 @@
+#include <c10/macros/Export.h>
```
Why not just put this in `torch/csrc/profiler/iit.cpp`?

jingxu10 replied:
Because `TORCH_API` is required in `itt_wrapper.cpp`, which was defined in `c10/macros/Export.h`.

robieta commented on torch/csrc/itt.cpp:
```diff
@@ -0,0 +1,13 @@
+#include <torch/csrc/utils/pybind.h>
```
I don't see why this shouldn't be inlined.

jingxu10 replied:
This is to bind cpp function to python function in this file. Any suggestion how to deal with it?

malfet commented:
Please rebase against `viable/strict` and if CI is green it's good to go

jingxu10 commented:
Hi @malfet , I rebased against `viable/strict`.

malfet commented:
@pytorchbot rebase

malfet reviewed (changes requested):
After rebase it fails lintrunner constraints, please fix.

jingxu10 commented:
@malfet  Every time when I make changes to the PR, I get this "workflows awaiting approval" issue (First-time contributors need a maintainer to approve running workflows.). Is there anything wrong with my account privilege? or is this expected? I cannot trigger CI after updating the PR.

jingxu10 commented:
Hi @malfet , Failed CIs are not related to code changes in the PR.

malfet commented:
@pytorchbot merge -f

malfet commented:
@pytorchbot revert -m "broke trunk, see https://hud.pytorch.org/pytorch/pytorch/commit/f988aa2b3ff77d5aa010bdaae4e52c6ee345c04d " -c ignoredsignal

malfet commented:
>  Failed CIs are not related to code changes in the PR.
Hmm, no they are related, it failed public API tests across the board

jingxu10 commented:
The failure error shows as the following (https://hud.pytorch.org/pytorch/pytorch/commit/f988aa2b3ff77d5aa010bdaae4e52c6ee345c04d), but I don't find in my code changes that could cause this naming issue to pytorch modules. Any idea how to can debug this?
```...
2022-06-30T07:15:22.2177153Z # torch.utils.model_dump.Dict:
2022-06-30T07:15:22.2178130Z   - Is public: it is an attribute that does not start with `_` on a module that does not have `__all__` defined
2022-06-30T07:15:22.2179551Z   - Does NOT look public: because its `__module__` attribute (`typing`) is not within the torch library or does not start with the submodule where it is defined (`torch.utils.model_dump`)
2022-06-30T07:15:22.2180682Z   - You can do either of these two things to fix this problem:
2022-06-30T07:15:22.2181891Z     - To make it NOT public: either define a `__all__` for `torch.utils.model_dump` or add a `_` at the beginning of the name
2022-06-30T07:15:22.2183155Z     - To make it look public: make sure the `__module__` is properly set and points to a submodule of `torch.utils.model_dump`
2022-06-30T07:15:22.2184007Z # torch.utils.tensorboard.RecordWriter:
2022-06-30T07:15:22.2185337Z   - Is public: it is an attribute that does not start with `_` on a module that does not have `__all__` defined
2022-06-30T07:15:22.2186919Z   - Does NOT look public: because its `__module__` attribute (`tensorboard.summary.writer.record_writer`) is not within the torch library or does not start with the submodule where it is defined (`torch.utils.tensorboard`)
2022-06-30T07:15:22.2188178Z   - You can do either of these two things to fix this problem:
2022-06-30T07:15:22.2189268Z     - To make it NOT public: either define a `__all__` for `torch.utils.tensorboard` or add a `_` at the beginning of the name
2022-06-30T07:15:22.2190466Z     - To make it look public: make sure the `__module__` is properly set and points to a submodule of `torch.utils.tensorboard`
2022-06-30T07:15:22.2191001Z 
2022-06-30T07:15:22.2191482Z ----------------------------------------------------------------------
2022-06-30T07:15:22.2192041Z Ran 5 tests in 3.286s
2022-06-30T07:15:22.2192314Z 
2022-06-30T07:15:22.2192565Z FAILED (failures=1, expected failures=3)
2022-06-30T07:15:22.2192902Z 
2022-06-30T07:15:22.2193085Z Generating XML reports...
2022-06-30T07:15:22.2194102Z Generated XML report: test-reports/python-unittest/test_public_bindings/TEST-TestPublicBindings-20220630071518.xml
2022-06-30T07:15:22.5898450Z Traceback (most recent call last):
2022-06-30T07:15:22.5898828Z   File "test/run_test.py", line 945, in <module>
2022-06-30T07:15:22.5902480Z     main()
2022-06-30T07:15:22.5902869Z   File "test/run_test.py", line 923, in main
2022-06-30T07:15:22.5906011Z     raise RuntimeError(err_message)
2022-06-30T07:15:22.5906358Z RuntimeError: test_public_bindings failed!
```

jingxu10 commented:
@malfet I think I figured out the root cause, and updated the code. Could you reinit the CI?
I added `range` into `__all__` of the `torch/profiler/itt.py`.

malfet commented:
@jingxu10 have you push new changes to the branch already?

jingxu10 commented:
yes, updated the branch already.

jingxu10 commented:
Hi @malfet , finished, could you please take a look at the CI results?

jingxu10 commented:
Hi @malfet , rebased to the latest code base, and CIs are all green.

malfet commented:
@pytorchbot merge
